apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -0,0 +1,174 @@
+        ///     unset the default will be used.</param>
+        public static void AddGoogleTrace(
+            this IServiceCollection services, string projectId,
+            TraceConfiguration config = null, Task<TraceServiceClient> clientTask = null)
+        {
+            GaxPreconditions.CheckNotNull(services, nameof(services));
+            GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+
+            clientTask = clientTask ?? TraceServiceClient.CreateAsync();
+            config = config ?? TraceConfiguration.Create();
+
+            IConsumer<TraceProto> consumer = ConsumerFactory<TraceProto>.GetConsumer(
+                 new GrpcTraceConsumer(clientTask), TraceSizer.Instance, config.BufferOptions);
+
+            services.AddSingleton(new ProjectId { Id = projectId });
+            services.AddSingleton(consumer);
+            services.AddSingleton(TraceIdFactory.Create());
+            services.AddSingleton(RateLimitingTraceOptionsFactory.Create(config));
+            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+
+            services.AddScoped(CreateTraceHeaderContext);
+            services.AddScoped(CreateManagedTracer);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// </summary>
+        internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
+        {
+            var accessor = provider.GetService<IHttpContextAccessor>();
+            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            return TraceHeaderContext.FromHeader(header);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IManagedTracer"/> based on the <see cref="TraceHeaderContext"/> and
+        /// the rate limiter.
+        /// </summary>
+        internal static IManagedTracer CreateManagedTracer(IServiceProvider provider)
+        {
+            var headerContext = provider.GetService<TraceHeaderContext>();
+            var rateLimitingFactory = provider.GetService<RateLimitingTraceOptionsFactory>();
+            var projectId = provider.GetService<ProjectId>();
+            var traceIdFactory = provider.GetService<TraceIdFactory>();
+            var consumer = provider.GetService<IConsumer<TraceProto>>();
+
+            // Check that we have all the needed services.
+            if (headerContext == null || rateLimitingFactory == null ||
+                projectId == null || traceIdFactory == null || consumer == null) {
+                throw new InvalidOperationException("Ensure Google Cloud Trace is properly set up.");
+            }
+
+            // If the trace header says to trace or if the rate limiter allows tracing continue.
+            if (!headerContext.ShouldTrace)
+            {
+                TraceOptions options = rateLimitingFactory.CreateOptions();
```

Comment:
I'm confused about the logic here. It looks like we'll return a no-op tracer only if the header context says we shouldn't trace *and* the rate limiter says we shouldn't trace. Shouldn't it be something like:

```csharp
if (!headerContext.ShouldTrace || !rateLimitingFactory.CreateOptions().ShouldTrace)
{
    return DoNothingTracer.Instance;
}
```
?